pkg/admission/serve.go

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/kubernetes/kubernetes/blob/5539a5b80fe45cc04df4f53bd1047cd6d998bcf1/test/images/agnhost/webhook/main.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Kubernetes Authors.
+
+package admission
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	v1 "k8s.io/api/admission/v1"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+)
+
+// admitv1beta1Func handles a v1beta1 admission
+type admitv1beta1Func func(context.Context, log.Logger, v1beta1.AdmissionReview, *kubernetes.Clientset) *v1beta1.AdmissionResponse
+
+// AdmitV1Func handles a v1 admission
+type AdmitV1Func func(context.Context, log.Logger, v1.AdmissionReview, *kubernetes.Clientset) *v1.AdmissionResponse
+
+func delegateV1beta1AdmitToV1(f AdmitV1Func) admitv1beta1Func {
+	return func(ctx context.Context, logger log.Logger, review v1beta1.AdmissionReview, api *kubernetes.Clientset) *v1beta1.AdmissionResponse {
+		in := v1.AdmissionReview{Request: convertAdmissionRequestToV1(review.Request)}
+		out := f(ctx, logger, in, api)
+		return convertAdmissionResponseToV1beta1(out)
+	}
+}
+
+// Serve handles the http portion of a request prior to handing to a V1Handler.
+func Serve(admit AdmitV1Func, logger log.Logger, api *kubernetes.Clientset) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var body []byte
+		if r.Body != nil {
+			if data, err := io.ReadAll(r.Body); err == nil {
+				body = data
+			}
+		}
+
+		// verify the content type is accurate
+		contentType := r.Header.Get("Content-Type")
+		if contentType != "application/json" {
+			level.Error(logger).Log("msg", "wrong content type", "contentType", contentType, "expected", "application/json")
+			return
```

Comment:
Should we return a 400 Bad Request here too?

Reply:
Makes sense, done in https://github.com/grafana/rollout-operator/pull/25/commits/a95ea040dc98c410c405198de4ee470721a9ede2